### ISSUE TITLE
Docker Containers do not find machine id from pyarmor-auth server

### ISSUE THREAD

**Issue by camgian** [dashingsoft/pyarmor#1534]
pyarmor-auth is running on the docker host 

(pyarmor)camgian@aiml-one:/camgian/etc/pyarmor$ pyarmor-auth pyarmor-device-regfile-6117.1.zip
2023-10-16 14:27:30,228: work path: /home/camgian/.pyarmor/docker
2023-10-16 14:27:30,229: register "pyarmor-device-regfile-6117.1.zip"
2023-10-16 14:27:30,235: machine id in group license: l971134fe71b1a329a6f22ca02df67844
2023-10-16 14:27:30,235: got machine id: l971134fe71b1a329a6f22ca02df67844
2023-10-16 14:27:30,235: this machine id matchs group license
2023-10-16 14:27:30,237: listen container auth request on 0.0.0.0:29092

docker container started with:

docker run -itd --add-host=host.docker.internal:host-gateway python bash

root@6d4671d089ae:/# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.2  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ac:11:00:02  txqueuelen 0  (Ethernet)
        RX packets 9382  bytes 13418595 (12.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 798  bytes 52297 (51.0 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

root@6d4671d089ae:/# ping 172.17.0.1
PING 172.17.0.1 (172.17.0.1) 56(84) bytes of data.
64 bytes from 172.17.0.1: icmp_seq=1 ttl=64 time=0.051 ms
64 bytes from 172.17.0.1: icmp_seq=2 ttl=64 time=0.060 ms
^C
--- 172.17.0.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1007ms
rtt min/avg/max/mdev = 0.051/0.055/0.060/0.004 ms

root@6d4671d089ae:/# pip install pyarmor.cli-8.3.11-py3-none-any.whl
Processing /pyarmor.cli-8.3.11-py3-none-any.whl
Requirement already satisfied: pyarmor.cli.core~=4.3.6 in /usr/local/lib/python3.12/site-packages (from pyarmor.cli==8.3.11) (4.3.6)
Installing collected packages: pyarmor.cli
Successfully installed pyarmor.cli-8.3.11
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

[notice] A new release of pip is available: 23.2.1 -> 23.3
[notice] To update, run: pip install --upgrade pip
root@6d4671d089ae:/# which conda
root@6d4671d089ae:/# ls
bin   dev  home  lib32  libx32  mnt  proc                pyarmor-device-regfile-6117.1.zip    root  sbin  sys  usr
boot  etc  lib   lib64  media   opt  pyarmor-8.3.11.zip  pyarmor.cli-8.3.11-py3-none-any.whl  run   srv   tmp  var
root@6d4671d089ae:/# pip install pyarmor-8.3.11.zip
Processing /pyarmor-8.3.11.zip
  Preparing metadata (setup.py) ... done
Requirement already satisfied: pyarmor.cli.core~=4.3.6 in /usr/local/lib/python3.12/site-packages (from pyarmor==8.3.11) (4.3.6)
Building wheels for collected packages: pyarmor
  Building wheel for pyarmor (setup.py) ... done
  Created wheel for pyarmor: filename=pyarmor-8.3.11-py3-none-any.whl size=2709951 sha256=769ace2a0d17f2b97cf7b34073df97d7055dd959199262570db9472283af76ed
  Stored in directory: /root/.cache/pip/wheels/c7/de/77/b6496bf14d83ad83410b8c6ad4b2508156ccbf1caa95b77597
Successfully built pyarmor
Installing collected packages: pyarmor
Successfully installed pyarmor-8.3.11
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

[notice] A new release of pip is available: 23.2.1 -> 23.3
[notice] To update, run: pip install --upgrade pip
root@6d4671d089ae:/# which pyarmor
/usr/local/bin/pyarmor
root@6d4671d089ae:/# pyarmor reg pyarmor-device-regfile-6117.1.zip
INFO     Python 3.12.0
INFO     Pyarmor 8.3.11 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register "pyarmor-device-regfile-6117.1.zip"
INFO     machine id in group license: l971134fe71b1a329a6f22ca02df67844
INFO     got machine id: lfcb7abb4e205e2869aa6f17d67f7ffb4
INFO     got machine id: ifcb7abb4e205e2869aa6f17d67f7ffb4
INFO     got machine id: k0e6eb566945da87cfd43525d8fb8bf0a
INFO     got machine id: g0e6eb566945da87cfd43525d8fb8bf0a
INFO     got machine id: bda8bfe4af2b0705a4761f58edb448243
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
root@6d4671d089ae:/#

**Comment by jondy** [comment]
Is there any log in console of `pyarmor-auth`?

And try to run `pyarmor -d reg xxx` in docker containter to check the log to find more information.

**Comment by goraje** [comment]
This is actually related to the [#1528](https://github.com/dashingsoft/pyarmor/issues/1528). Unfortunately the proposed fix is not working and is causing the exact error as mentioned here. I'll just paste what I wrote there:

Hi, unfortunately the proposed fix is not working. This is related directly to the patch that was merged to 8.3.11 [0b68e9ec7](https://github.com/dashingsoft/pyarmor/commit/0b68e9ec78139a1e9b00d1cc026d5f8c52bf1834). The conversion `bytes([x - 87 for x in MACHFLAGS])` is invalid because it is called on negative integers. This can be confirmed by running with the `--debug` flag.
```
INFO     Python 3.11.6
INFO     Pyarmor 8.3.11 (trial), 000000, non-profits
INFO     Platform linux.aarch64
DEBUG    native platform linux.aarch64
DEBUG    home path: /root/.pyarmor
INFO     register "pyarmor-device-regfile-6005.6.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     machine id in group license: l98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: l5bd8dfd8ad4f100902d43e6e0e8025be
INFO     got machine id: i5bd8dfd8ad4f100902d43e6e0e8025be
INFO     got machine id: k58444ccf00d89f5aa28e95778267b4d4
INFO     got machine id: g58444ccf00d89f5aa28e95778267b4d4
INFO     got machine id: b812ceb5191d569f7f09d136d173e5801
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
DEBUG    host.docker.internal:29092:bytes must be in range(0, 256)
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
```

**Comment by jondy** [comment]
Please upgrade Pyarmor to 8.4.0

Then generate new device file for docker host by 8.4.0

Make sure machine id doesn't change after reboot or started docker.

**Comment by goraje** [comment]
Hi, the changes introduced in 8.4.0 have made registration possible, however obfuscation inside the container after registration still seems to be broken.
```
root@4b279a129f6b:/# pyarmor reg pyarmor-device-regfile-6005.7.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.7.zip"
INFO     machine id in group license: m98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: m47f31be7ced2f9161c5221e56364fac2
INFO     got machine id: l47f31be7ced2f9161c5221e56364fac2
INFO     got machine id: i98444de277e1b3dd9b17d2181631a73c
INFO     got machine id: k1a298b2ce465500b355f6c9d777b845f
INFO     got machine id: g1a298b2ce465500b355f6c9d777b845f
INFO     got machine id: bb6889a725ecfcc7acd18b1dd249f4df9
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     got docker host machine id: m98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: l98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: i98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: kc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: gc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: b28c5a05ba0fa8e759f3e172660d31c04
INFO     This license registration information:

License Type    : pyarmor-group
License No.     : pyarmor-vax-006005
License To      : *****
License Product : ******

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation

root@4b279a129f6b:/# pyarmor gen foo.py
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (group), 006005, ******
INFO     Platform linux.aarch64
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
ERROR    invalid license token, try to run `pyarmor reg` to register license again
```

and `pyarmor.error.log` still showing:

```
266 MainProcess 2023-10-18 06:58:09,818
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 712, in main
    main_entry(sys.argv[1:])
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 700, in main_entry
    return args.func(ctx, args)
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 230, in cmd_gen
    builder.process(options)
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/generate.py", line 147, in process
    Pytransform3.pre_build(self.ctx)
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/core/__init__.py", line 117, in pre_build
    m = Pytransform3.init(ctx)
        ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/core/__init__.py", line 97, in init
    m.init_ctx(ctx)
RuntimeError: invalid license token, try to run `pyarmor reg` to register license again
```

**Comment by camgian** [comment]
Jondy,

When I try to upgrade using the regcode it fails to create  the "pyarmor-regfile-6117.zip" to generate the device file.

Thank you in advance for your help!

Clark

See below:

***@***.***:/camgian/etc/pyarmor$ pyarmor reg -u -p pyarmor-group  .regcode/pyarmor-regcode-6117.txt
INFO     Python 3.11.4
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64

Pyarmor 8 changes EULA and uses new commands
It's different from previous Pyarmor totally
Please read this import notes first:
https://github.com/dashingsoft/pyarmor/issues/980
Do not upgrade to Pyarmor 8 if don't know what are changed

I have known the changes of Pyarmor 8? (yes/no/help) yes
INFO     prepare ".regcode/pyarmor-regcode-6117.txt"
INFO     query key file from server
WARNING  this license is bind to product "Pyarmor-group"
WARNING  it can not be changed to "pyarmor-group"

You are about to upgrade old Pyarmor license to Pyarmor Basic License

The upgraded license information will be

License Type    : pyarmor-group
License To      : Camgian Corporation
License Product : Pyarmor-group

This license is about to be used for product "Pyarmor-group"

Are you sure to continue? (yes/no) yes
INFO     process activation file ".regcode/pyarmor-regcode-6117.txt"
INFO     send request to server
INFO     write registration file "pyarmor-regfile.zip"
INFO     register "pyarmor-regfile.zip"
ERROR    File is not a zip file
***@***.***:/camgian/etc/pyarmor$

***@***.***:/camgian/etc/pyarmor$ pyarmor -v
Pyarmor 8.4.0 (trial), 000000, non-profits

License Type    : pyarmor-trial
License No.     : pyarmor-vax-000000
License To      :
License Product : non-profits

BCC Mode        : No
RFT Mode        : No

Notes
* Can't obfuscate big script and mix str

***@***.***:/camgian/etc/pyarmor$


________________________________
From: Jondy ***@***.***>
Sent: Tuesday, October 17, 2023 11:09 PM
To: dashingsoft/pyarmor ***@***.***>
Cc: System Administration Team ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [dashingsoft/pyarmor] Docker Containers do not find machine id from pyarmor-auth server (Issue #1534)

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize and verify the sender and verify the content is safe.


Please upgrade Pyarmor to 8.4.0

Then generate new device file for docker host by 8.4.0

Make sure machine id doesn't change after reboot or started docker.

**Comment by jondy** [comment]
Hi guy, you need not upgrade regfile, but upgrading Pyarmor to 8.4.0

Please check the documentation `how-to` guide about using license

**Comment by camgian** [comment]
Hi Jondy,

This is not resolved, the inital registration of pyarmor is jacked up and therefore the DEVICE registration does not properly create the the device file.  I have gone so far as to purchase a second license to  run run the INITIAL registration so we now own 6058 and 6117 and neither will register a  device file.


***@***.***:/camgian/etc/pyarmor$ pyarmor reg -g 1 regfile/pyarmor-regfile-6058.zip
INFO     Python 3.11.4
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register device file ".pyarmor/group/pyarmor-group-device.1"
INFO     use group license "regfile/pyarmor-regfile-6058.zip"
INFO     send request to server
ERROR    device (1) is not for this machine

***@***.***:/camgian/etc/pyarmor$ pyarmor reg -g 1 regfile/pyarmor-regfile-6117.zip
INFO     Python 3.11.4
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register device file ".pyarmor/group/pyarmor-group-device.1"
INFO     use group license "regfile/pyarmor-regfile-6117.zip"
INFO     send request to server
ERROR    device (1) is not for this machine

Or will they show the main server as being registered.

***@***.***:/camgian/etc/pyarmor$ pyarmor -v
Pyarmor 8.4.0 (trial), 000000, non-profits

License Type    : pyarmor-trial
License No.     : pyarmor-vax-000000
License To      :
License Product : non-profits

BCC Mode        : No
RFT Mode        : No

Notes
* Can't obfuscate big script and mix str

Could we get the INITIAL registration corrected on both of the licenses.

Thank You in advance for your help.

Clark


***@***.***:/camgian/etc/pyarmor$ conda activate /camgian/conda_envs/pyarmor
***@***.***:/camgian/etc/pyarmor$ python --version
Python 3.11.4
***@***.***:/camgian/etc/pyarmor$ pyarmor -v
Pyarmor 8.4.0 (trial), 000000, non-profits

License Type    : pyarmor-trial
License No.     : pyarmor-vax-000000
License To      :
License Product : non-profits

BCC Mode        : No
RFT Mode        : No

Notes
* Can't obfuscate big script and mix str

***@***.***:/camgian/etc/pyarmor$ pyarmor reg -g 1 regfile/pyarmor-regfile-6058.zip
INFO     Python 3.11.4
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register device file ".pyarmor/group/pyarmor-group-device.1"
INFO     use group license "regfile/pyarmor-regfile-6058.zip"
INFO     send request to server
ERROR    device (1) is not for this machine
***@***.***:/camgian/etc/pyarmor$ pyarmor reg -g 1 regfile/pyarmor-regfile-6117.zip
INFO     Python 3.11.4
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register device file ".pyarmor/group/pyarmor-group-device.1"
INFO     use group license "regfile/pyarmor-regfile-6117.zip"
INFO     send request to server
ERROR    device (1) is not for this machine
***@***.***:/camgian/etc/pyarmor$ ls -alrt
total 8
drwxrwxr-x 1 camgian camgian  164 Oct 12 19:46 ..
-rw-rw-r-- 1 camgian camgian 2658 Oct 16 13:09 README_PYARMOR_INITIAL_REGISTRATION
-rw-rw-r-- 1 camgian camgian 3539 Oct 16 13:43 README_PYARMOR_GROUP_REGISTRATION
drwxrwxr-x 1 camgian camgian   96 Oct 16 18:27 .regcode
drwxrwxr-x 1 camgian camgian  254 Oct 16 22:18 docker-registration
drwxrwxr-x 1 camgian camgian   96 Oct 18 15:04 regfile
drwxrwxr-x 1 camgian camgian  106 Oct 18 15:07 .pyarmor
drwxrwxr-x 1 camgian camgian  220 Oct 19 13:54 .
***@***.***:/camgian/etc/pyarmor$ ls .regcode/
pyarmor-regcode-6058.txt  pyarmor-regcode-6117.txt
***@***.***:/camgian/etc/pyarmor$ ls -alrt regfile/
total 16
-rw-rw-r-- 1 camgian camgian 4600 Oct 16 13:04 pyarmor-regfile-6117.zip
drwxrwxr-x 1 camgian camgian   96 Oct 18 15:04 .
-rw-rw-r-- 1 camgian camgian 4598 Oct 19 13:52 pyarmor-regfile-6058.zip
drwxrwxr-x 1 camgian camgian  220 Oct 19 13:54 ..
***@***.***:/camgian/etc/pyarmor$



________________________________
From: Jondy ***@***.***>
Sent: Wednesday, October 18, 2023 6:34 PM
To: dashingsoft/pyarmor ***@***.***>
Cc: System Administration Team ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [dashingsoft/pyarmor] Docker Containers do not find machine id from pyarmor-auth server (Issue #1534)

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize and verify the sender and verify the content is safe.


Closed #1534<https://github.com/dashingsoft/pyarmor/issues/1534> as completed.
